cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/visitor/DataRequirementsVisitor.java

```diff
@@ -0,0 +1,246 @@
+            throw new RuntimeException(translator.getErrors().get(0).getMessage());
+        }
+        return translator;
+    }
+
+    protected static LibrarySourceProvider buildLibrarySource(Repository repository) {
+        AdapterFactory adapterFactory = AdapterFactory.forFhirContext(repository.fhirContext());
+        return new RepositoryFhirLibrarySourceProvider(
+                repository, adapterFactory, new LibraryVersionSelector(adapterFactory));
+    }
+
+    protected LibraryManager createLibraryManager(Repository repository) {
+        var librarySourceProvider = buildLibrarySource(repository);
+        var sourceProviders = new ArrayList<>(Arrays.asList(librarySourceProvider, librarySourceProvider));
+        var modelManager = evaluationSettings.getModelCache() != null
+                ? new ModelManager(evaluationSettings.getModelCache())
+                : new ModelManager();
+        var libraryManager = new LibraryManager(
+                modelManager,
+                evaluationSettings.getCqlOptions().getCqlCompilerOptions(),
+                evaluationSettings.getLibraryCache());
+        libraryManager.getLibrarySourceLoader().clearProviders();
+        for (var provider : sourceProviders) {
+            libraryManager.getLibrarySourceLoader().registerProvider(provider);
+        }
+        return libraryManager;
+    }
+
+    protected <T extends ICompositeType & IBaseHasExtensions> void recursiveGather(
+            IDomainResource resource,
+            List<String> gatheredResources,
+            List<T> relatedArtifacts,
+            Repository repository,
+            List<String> capability,
+            List<String> include,
+            List<String> artifactVersion,
+            List<String> checkArtifactVersion,
+            List<String> forceArtifactVersion)
+            throws PreconditionFailedException {
+        if (resource != null && !gatheredResources.contains(resource.getId())) {
+            gatheredResources.add(resource.getId());
+            var fhirVersion = resource.getStructureFhirVersionEnum();
+            var adapter = AdapterFactory.forFhirVersion(fhirVersion).createKnowledgeArtifactAdapter(resource);
+            findUnsupportedCapability(adapter, capability);
+            processCanonicals(adapter, artifactVersion, checkArtifactVersion, forceArtifactVersion);
+            var reference = adapter.hasVersion()
+                    ? adapter.getUrl().concat(String.format("|%s", adapter.getVersion()))
+                    : adapter.getUrl();
+            boolean addArtifact = relatedArtifacts.stream()
+                    .noneMatch(ra -> KnowledgeArtifactAdapter.getRelatedArtifactReference(ra)
+                            .equals(reference));
+            if (addArtifact) {
+                relatedArtifacts.add(KnowledgeArtifactAdapter.newRelatedArtifact(
+                        fhirVersion, "depends-on", reference, adapter.getDescriptor()));
+            }
```

Comment:
In the event that an unversioned `relatedArtifact` reference already exists, this would result in a duplicate `relatedArtifact` with a reference?

Reply:
resolving as per our discussion 👍🏾 (TLDR : not possible for there to be unversioned `relatedArtifact`s)